subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/AnnotationProcessorDetectionIntegrationTest.groovy

```diff
@@ -0,0 +1,186 @@
+
+        when:
+        // Modify main source file.  Would normally be incremental.
+        file('app/src/main/java/App.java').text = "@Incremental class App { /*change*/ }"
+        succeeds 'app:compileJava'
+
+        then:
+        skipped ':inc:compileJava'
+
+        // Verify that it was incremental (B not recompiled).
+        outputs.recompiledClasses 'App', 'AppIncremental'
+    }
+
+    def "presence of non-incremental AP should force non-incremental build"() {
+        appWithBothProcessors()
+        outputs.snapshot { run "compileJava" }
+
+        when:
+        succeeds 'app:compileJava'
+
+        then:
+        file('app/build/generated-sources/AppIncremental.java').exists()
+        file('app/build/generated-sources/AppNonIncremental.java').exists()
+
+        when:
+        // Trivial reformatting of the main source file.  Would normally be incremental.
+        file('app/src/main/java/App.java').text = "@Incremental @NonIncremental class App { /*change*/ }"
+        succeeds 'app:compileJava'
+
+        then:
+        skipped ':inc:compileJava'
+        skipped ':noninc:compileJava'
+
+        // Verify app was rebuilt non-incrementally.
+        outputs.recompiledClasses 'App', 'B', 'AppNonIncremental', 'AppIncremental'
+    }
+
+    def "deleting an annotation should delete its generated file"() {
+        appWithBothProcessors()
+        outputs.snapshot { run "compileJava" }
+
+        when:
+        file('app/src/main/java/App.java').text = "@Incremental class App { /*removed annotation*/ }"
+        succeeds 'app:compileJava'
+
+        then:
+        outputs.deletedClasses 'AppNonIncremental'
+    }
+
+    private void createBothProcessors() {
+        file('settings.gradle') << "include 'inc'\n"
+        new IncrementalAnnotationProcessorFixture().writeLibraryTo(file('inc'))
+        file('settings.gradle') << "include 'noninc'\n"
+        new NonIncrementalAnnotationProcessorFixture().writeLibraryTo(file('noninc'))
+    }
+
+    private void appWithBothProcessors() {
+        incrementalBuild()
+
+        subproject('app') {
+            'build.gradle'("""
+                ${buildDotGradleSansDeps()}
+                dependencies {
+                  compile project(':inc')
+                  compile project(':noninc')
+                  annotationProcessor project(':inc')
+                  annotationProcessor project(':noninc')
+                }
+            """)
+            src {
+                main {
+                    java {
+                        'App.java'("@Incremental @NonIncremental class App {}")
+                        'B.java'("class B {}")
+                    }
+                }
+            }
+        }
+    }
+
+    private void appWithIncapProcessorOnly() {
+        incrementalBuild()
+
+        subproject('app') {
+            'build.gradle'("""
+                ${buildDotGradleSansDeps()}
+                dependencies {
+                  compile project(':inc')
```

Comment:
⭕️ `compileOnly` - it's just there for annotations, we don't want to leak the processor implementation to our downstream clients. Good citizens among the processor authors separate their annotations and processor into two jars.

Reply:
Done.